core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java

```diff
@@ -1024,6 +1025,24 @@ protected Object visitBooleanLiteral(BooleanLiteral node, Object context)
         @Override
         protected Object visitFunctionCall(FunctionCall node, Object context)
         {
+            ResolvedFunction resolvedFunction = metadata.decodeFunction(node.getName());
+
+            if (optimize && resolvedFunction.getSignature().getName().equals(TryFunction.NAME) &&
```

Comment:
Instead of adding special handling for `try`, I'd expand the code below to support functions with lambda arguments.

Reply:
I did some investigation and we actually support evaluate for functions with lambda arguments. 
`try(cast(20230303 as VARCHAR))` cannot be optimized because `tryFunction` is marked as a non-deterministic function. But in fact, whether try is deterministic or not depends on its argument (`LambdaFunctionInterface`) and whether it has side effects during execution. It looks like we mark all functions with Lambda arguments as non-deterministic, just in case. 

Perhaps deterministic in addition to true / false we need to provide a third value, based on the Lambda argument. Looking forward to your opinion and if you support mine, I will put it into practice. @dain

Reply:
Indeed. The current logic to decide whether an expression is deterministic is rather conservative (at the risk of false positives). Additionally, it doesn't properly handle higher-order functions because we introduced that concept before lambda expressions were added to the engine.

Making it more precise would be a good thing.